### PR TITLE
docs: update readme, docs layout and add user, dev docs [skip ci]

### PR DESCRIPTION
**Pull request by leelavg**
- Updated README with deployment instructions
- Created layout for user and dev guides
- Renamed dir `doc` to `docs`
- Added user guide
- Added development guide

Signed-off-by: Leela Venkaiah G <lgangava@redhat.com>

### CODE REVIEW

File: docs/dev-guide/topolvm-csi.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# Topolvm CSI
+
+- LVM Operator orchestrates Topolvm CSI by creating and managing all required
+  resources for seamless usage of Logical Volumes as Persistent Volumes
+- This document outlines only the changes made to deployments of Topolvm CSI
+  components and further info please consult topolvm [docs][topolvm-docs]
```
Comment (nbalacha) [inline docs/dev-guide/topolvm-csi.md:6]:
```suggestion
- Please refer to topolvm [docs][topolvm-docs] for more details on topolvm
```

File: docs/dev-guide/topolvm-csi.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+## CSI Driver
+
+- *csiDriver* reconcile unit deploys Topolvm CSIDriver resource
+- Ephemeral volumes support is removed (not in the code) due to below reasons:
```
Comment (nbalacha) [inline docs/dev-guide/topolvm-csi.md:11]: The LVM operator does not expose support for ephemeral volumes.

Please remove the section on reasons.

File: docs/dev-guide/topolvm-csi.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+     LVMO will try to include that support
+  2. Technical difficulty on deciding default device class when there are
+     multiple volume groups across nodes
+
+## Topolvm Controller
+
+- *topolvmController* reconcile unit deploys a single Topolvm Controller plugin
+  deployment and manages any updates to the deployment
+- Topolvm scheduler is not used for pod scheduling but only CSI StorageCapacity
+  tracking feature is used
```
Comment (nbalacha) [inline docs/dev-guide/topolvm-csi.md:23]:
```suggestion
- Topolvm scheduler is not used for pod scheduling. The CSI StorageCapacity
  feature is used by the scheduler.
```

File: docs/dev-guide/topolvm-csi.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+- An init container generates openssl certs to be used in topolvm controller
+  which will be soon replaced with cert-manager
+
+## Topolvm Node and LVMd
+
+- *topolvmNode* reconcile unit deploys and manages topolvm node plugin and lvmd
+  daemonset and scales it based on node selector on which volume groups exists
```
Comment (nbalacha) [inline docs/dev-guide/topolvm-csi.md:30]:
```suggestion
  daemonset based on the node selector specified in the deviceClasses in LVMCluster.
```

File: docs/dev-guide/topolvm-csi.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+- An init container polls for the availability of lvmd config file before start
+  of lvmd and topolvm-node containers
+
```
Comment (nbalacha) [inline docs/dev-guide/topolvm-csi.md:33]:
```suggestion
- An init container polls for the availability of lvmd config file before starting
  the lvmd and topolvm-node containers

```

File: docs/user-guide/README.md [hunk 6]
```diff
@@ -0,0 +1,344 @@
+## Deploying LVMCluster CR
+
+This guide assumes you followed steps in [Readme][repo-readme] and LVM operator
+(hereafter LVMO) is running in your cluster.
+
+Below are the available disks in one of the kubernetes nodes and there are no
```
Comment (nbalacha) [inline docs/user-guide/README.md:6]:
```suggestion
Below are the available disks in our test kubernetes cluster node and there are no
```

File: docs/user-guide/README.md [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+existing Physical Volumes, Volume Groups and Logical Volumes
```
Comment (nbalacha) [inline docs/user-guide/README.md:7]:
```suggestion
existing LVM Physical Volumes, Volume Groups and Logical Volumes
```

File: docs/user-guide/README.md [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+
+``` console
+kubectl get pods
+NAME                                  READY   STATUS    RESTARTS   AGE
+controller-manager-8bf864c85-8zjlp    3/3     Running   0          31m
+topolvm-controller-694bc98b87-x6dxn   4/4     Running   0          10m
+topolvm-node-pbcth                    4/4     Running   0          10m
+vg-manager-f979f                      1/1     Running   0          10m
+```
+Most notable changes by end of above operation is creation of physical volumes
```
Comment (nbalacha) [inline docs/user-guide/README.md:72]:
```suggestion
When all pods are ready, the LVM physical volumes and volume groups should be created on the host.
```

File: docs/user-guide/README.md [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+  /dev/nvme1n1 vg1 lvm2 a--  <1.46t <1.46t
+  /dev/nvme2n1 vg1 lvm2 a--  <1.46t <1.46t
+sh-4.4# vgs
+  VG  #PV #LV #SN Attr   VSize  VFree 
+  vg1   3   0   0 wz--n- <4.37t <4.37t
+```
+
+Before moving further we can look for other resources apart from above pods
+created by LVMO after deploying LVMCluster CR
+
```
Comment (nbalacha) [inline docs/user-guide/README.md:88]:
```suggestion
Confirm that the following resources have been created on the cluster:

```

File: docs/user-guide/README.md [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+- When LVMCluster CR is deployed, which is expected to be a single entity
```
Comment (nbalacha) [inline docs/user-guide/README.md:89]: This point can be removed.

File: docs/user-guide/README.md [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+  containing details of all volume groups that needs to be created across
+  multiple nodes, two supporting Custom Resources are created by LVMO
+- LVMVolumeGroup: CR created and managed by LVMO which tracks individual volume
+  groups across multiple nodes
+- LVMVolumeGroupNodeStatus: CR created and managed by LVMO which tracks node status
```
Comment (nbalacha) [inline docs/user-guide/README.md:94]: Reword this section to show the actual resources that are created.

File: docs/user-guide/README.md [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+Note:
+- Reconciling multiple LVMCluster CRs by LVMO is not supported and erroneous
+- Custom resources LVMVolumeGroup and LVMVolumeGroupNodeStatus are managed by
+  LVMO and if need may arise users are expected to **ONLY** edit LVMCluster CR
+
+## Deploying PVC and App Pod
+
+- A successful reconciliation of LVMCluster CR will setup all the underlying
+  resources needed to be able to create a PVC and use it in app pod
+- Create a PVC as below and as the storage class created by LVMO is of
```
Comment (nbalacha) [inline docs/user-guide/README.md:113]:
```suggestion
- Create a PVC using the StorageClass created for the deviceClass. The PVC will not be bound until a pod claims the storage as the volume binding mode is set to WaitForFirstConsumer.
```

File: docs/user-guide/README.md [hunk 13]
```diff
@@ -0,0 +1,344 @@
...
+kubectl get pod | grep -E '^(app|NAME)'
+NAME                                  READY   STATUS    RESTARTS   AGE
+app-file                              1/1     Running   0          2m6s
+
+kubectl get pvc
+NAME           STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
+lvm-file-pvc   Bound    pvc-04a4f5f7-8665-4008-818a-490503c859f5   5Gi        RWO            topolvm-vg1    5m55s
+```
+- In a similar way, PVC with `volumeMode: Block` can also be created and
+  claimed by App pod as below
```
Comment (nbalacha) [inline docs/user-guide/README.md:174]:
```suggestion
- To create a PVC with `volumeMode: Block` , do the following:
```

File: docs/user-guide/README.md [hunk 14]
```diff
@@ -0,0 +1,344 @@
...
+    volumeDevices:
+    - devicePath: "/dev/xvda"
+      name: lvm-block-pvc
+  volumes:
+    - name: lvm-block-pvc
+      persistentVolumeClaim:
+        claimName: lvm-block-pvc
+EOF
+```
+- Status of both PVCs and App pods can be referred as below
```
Comment (nbalacha) [inline docs/user-guide/README.md:210]:
```suggestion
- Check the status of both PVCs and App pods:
```

File: docs/user-guide/README.md [hunk 15]
```diff
@@ -0,0 +1,344 @@
...
+NAME            STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
+lvm-block-pvc   Bound    pvc-56fe959a-1184-4fa6-9d66-b69c06d43cf4   5Gi        RWO            topolvm-vg1    104s
+lvm-file-pvc    Bound    pvc-04a4f5f7-8665-4008-818a-490503c859f5   5Gi        RWO            topolvm-vg1    13m
+```
+
+### Monitoring and Metrics
+
+- LVMO currently surfaces only Topolvm metrics and those can be viewed either
+  from UI or port-forwarding the service
+- Make sure the namespace LVMO is running in has below label or patch the
```
Comment (nbalacha) [inline docs/user-guide/README.md:227]:
```suggestion
- On Openshift clusters, set the label  "openshift.io/cluster-monitoring" on the namespace the LVMO is running in.
```

File: docs/user-guide/README.md [hunk 16]
```diff
@@ -0,0 +1,344 @@
...
+curl -s 127.0.0.1:41685/metrics | grep -Ei 'topolvm_volumegroup_.*?_bytes\{'
+topolvm_volumegroup_available_bytes{device_class="vg1",node="rhocs-bm4.lab.eng.blr.redhat.com"} 4.790222323712e+12
+topolvm_volumegroup_size_bytes{device_class="vg1",node="rhocs-bm4.lab.eng.blr.redhat.com"} 4.800959741952e+12
+```
+
+## Cleanup
+
+- Feature to auto cleanup volume groups after removing LVMCluster CR is coming
+  soon
+- Until then, please follow below procedure to cleanup left out resources
```
Comment (nbalacha) [inline docs/user-guide/README.md:250]:
```suggestion
- Until then, please do the following after deleting the LVMCluster CR to cleanup the resources created by the operator
```

File: docs/user-guide/README.md [hunk 17]
```diff
@@ -0,0 +1,344 @@
...
+  during CR deletion
+
+### Steps
+
+1. Make sure no PVC is bound to an App pod and remove the PVCs, all app pods can be removed in parallel and corresponding PVCs can be removed after pod deletion, below deletion is only for representation
```
Comment (nbalacha) [inline docs/user-guide/README.md:255]: Simplify the sentence. Suggestion:
Remove all apps using PVCs created with topolvm.

File: docs/user-guide/README.md [hunk 18]
```diff
@@ -0,0 +1,344 @@
...
+``` console
+# delete App pod first
+kubectl delete pod app-file
+pod "app-file" deleted
+
+# delete PVC which was used by above App pod
+kubectl delete pvc lvm-file-pvc
+persistentvolumeclaim "lvm-file-pvc" deleted
+
+# likewise remove for rest of App pods and corresponding PVCs
```
Comment (nbalacha) [inline docs/user-guide/README.md:265]: Delete this sentence and club all the deletions in one point.

File: docs/user-guide/README.md [hunk 19]
```diff
@@ -0,0 +1,344 @@
...
+
+kubectl delete pvc lvm-block-pvc
+persistentvolumeclaim "lvm-block-pvc" deleted
+```
+2. Make sure there are no `logicalvolumes` CRs which were created by topolvm
+``` console
+kubectl get logicalvolumes.topolvm.cybozu.com 
+No resources found
+```
+3. Take a json dump of LVMCluster CR contents which helps in finding out Volume
```
Comment (nbalacha) [inline docs/user-guide/README.md:277]:
```suggestion
3. Take a json dump of LVMCluster CR contents. It has the list of VGs and PVs created on the nodes.
```

File: docs/user-guide/README.md [hunk 20]
```diff
@@ -0,0 +1,344 @@
...
+   except operator (contorller-manager pod)
+``` console
+kubectl delete lvmclusters.lvm.topolvm.io lvmcluster-sample
+lvmcluster.lvm.topolvm.io "lvmcluster-sample" deleted
+
+kubectl get po
+NAME                                 READY   STATUS    RESTARTS   AGE
+controller-manager-8bf864c85-8zjlp   3/3     Running   0          125m
+```
+6. Login to the node and remove above noted volume groups and physical volumes
```
Comment (nbalacha) [inline docs/user-guide/README.md:312]:
```suggestion
6. Login to the node and remove the LVM volume groups and physical volumes
```

File: docs/user-guide/README.md [hunk 21]
```diff
@@ -0,0 +1,344 @@
...
+  WARNING: File locking is disabled.
+  Volume group "vg1" successfully removed
+sh-4.4# pvremove  /dev/nvme0n1 /dev/nvme1n1 /dev/nvme2n1 --nolock
+  WARNING: File locking is disabled.
+  Labels on physical volume "/dev/nvme0n1" successfully wiped.
+  Labels on physical volume "/dev/nvme1n1" successfully wiped.
+  Labels on physical volume "/dev/nvme2n1" successfully wiped.
+```
+7. `vg-manager` pod creates lvmd config file to be used in `topolvm-node` pod
+   during initialization and please remove that file as well from node
```
Comment (nbalacha) [inline docs/user-guide/README.md:324]:
```suggestion
7. Remove the lvmd.yaml config file from the node
```

File: README.md [hunk 22]
```diff
@@ -1,5 +1,100 @@
-# lvm-operator
-Operator that manages Topolvm
+# LVM Operator
 
-# Contents
-- [docs/design](doc/design/)
+LVM Operator is a local storage orchestrator for Kubernetes which manages
```
Comment (nbalacha) [inline README.md:3]: Please reword to The LVM Operator provides the ability to dynamically provision local storage via the topolvm CSI plugin.

File: README.md [hunk 23]
```diff
@@ -1,5 +1,100 @@
...
+[Topolvm][topolvm_repo]. It creates Logical Volume Groups, deploys Topolvm
```
Comment (nbalacha) [inline README.md:4]: It create LVM volume groups...

File: README.md [hunk 24]
```diff
@@ -1,5 +1,100 @@
...
+  be deployed
+```
+kubectl create -f config/samples/lvm_v1alpha1_lvmcluster.yaml -n lvm-operator-system
+```
+- Note: Resources are deployed in `lvm-operator-system` namespace when deployed
+  via manifests
+
+## Documentation
+
+- Please refer [docs][docs] directory for user and development guides.
```
Comment (nbalacha) [inline README.md:91]:
```suggestion
- Please refer to the [docs][docs] directory for user and development guides.
```

File: docs/dev-guide/custom-resource-definitions.md [hunk 25]
```diff
@@ -0,0 +1,126 @@
+This page outlines `spec` and `status` fields of all the CRs managed by LVMO
+
+## 1. LVMCluster
+
+- LVMCluster CR represents the volume groups that should be created and managed
+  across nodes with custom node selector, toleration and device selectors
+- Should be created and edited by user in operator installed namespace and only
+  single CR is supported, creation of multiple CRs is erroneous
```
Comment (nbalacha) [inline docs/dev-guide/custom-resource-definitions.md:8]:
```suggestion
- Should be created and edited by user in operator installed namespace.
- only a single instance is supported at present.
```

File: docs/dev-guide/custom-resource-definitions.md [hunk 26]
```diff
@@ -0,0 +1,126 @@
...
+| Field    | Type             | Description                                                      |
+|----------|------------------|------------------------------------------------------------------|
+| `spec`   | LVMClusterSpec   | Desired behaviour of LVM Cluster with Volume Groups across Nodes |
+| `status` | LVMClusterStatus | Observed status of LVM Cluster with Volume Groups across Nodes   |
+
+### LVMClusterSpec
+
+| Field           | Type                | Description                                                                        |
+|-----------------|---------------------|------------------------------------------------------------------------------------|
+| `toleraions`    | []corev1.Toleration | List of tolerations that Topolvm Controller, Node Plugin and VG Manager can run on |
```
Comment (nbalacha) [inline docs/dev-guide/custom-resource-definitions.md:19]: Typo: tolerations

Comment (nbalacha) [inline docs/dev-guide/custom-resource-definitions.md:19]:
```suggestion
| `tolerations`    | []corev1.Toleration | List of tolerations for the nodes that Topolvm Controller, Node Plugin and VG Manager can run on |
```

File: docs/dev-guide/custom-resource-definitions.md [hunk 27]
```diff
@@ -0,0 +1,126 @@
...
+| `deviceClasses` | []DeviceClass       | List of deviceclasses                                                              |
+
+#### DeviceClass
+
+| Field            | Type                  | Description                           |
+|------------------|-----------------------|---------------------------------------|
+| `name`           | string                | Name of the Volume Group              |
+| `deviceSelector` | \*DeviceSelector      | Selector to match devices on the node |
```
Comment (nbalacha) [inline docs/dev-guide/custom-resource-definitions.md:27]:
```suggestion
| `deviceSelector` | \*DeviceSelector      | Selector to match devices to be used |
```

File: docs/dev-guide/custom-resource-definitions.md [hunk 28]
```diff
@@ -0,0 +1,126 @@
...
+|-----------------------|---------------------|---------------------------------------|
+| `ready`               | bool                | State of the LVMCluster               |
+| `deviceClassStatuses` | []DeviceClassStatus | List of observed deviceClass statuses |
+
+#### DeviceClassStatus
+
+| Field        | Type         | Description                                          |
+|--------------|--------------|------------------------------------------------------|
+| `name`       | string       | Name of the device class                             |
+| `nodeStatus` | []NodeStatus | Status of the node that belongs to this device class |
```
Comment (nbalacha) [inline docs/dev-guide/custom-resource-definitions.md:46]:
```suggestion
| `nodeStatus` | []NodeStatus | Status of the VG on the nodes |
```

File: docs/dev-guide/custom-resource-definitions.md [hunk 29]
```diff
@@ -0,0 +1,126 @@
...
+
+#### NodeStatus
+
+| Field     | Type         | Description                                                           |
+|-----------|--------------|-----------------------------------------------------------------------|
+| `Node`    | string       | Name of the node                                                      |
+| `status`  | VGStatusType | One of `Ready`, `Failed`, `Degraded`                                  |
+| `devices` | []string     | List of Physical Volumes (devices) that are part of this device class |
+
+### Lifecycle
```
Comment (nbalacha) [inline docs/dev-guide/custom-resource-definitions.md:56]: Move this out of this file and to the deployment and user guides.

File: docs/dev-guide/custom-resource-definitions.md [hunk 30]
```diff
@@ -0,0 +1,126 @@
...
+
+1. Only a single volume group containing all available disks across schedulable
+   nodes is supported and implementation respecting tolerations, device andi
+   node selector fields is coming soon
+2. All fields in `status` is updated based on the status of volume groups
```
Comment (nbalacha) [inline docs/dev-guide/custom-resource-definitions.md:61]:
```suggestion
2. All fields in `status` are updated based on the status of volume groups
```

File: docs/dev-guide/custom-resource-definitions.md [hunk 31]
```diff
@@ -0,0 +1,126 @@
...
+   creation across nodes
+3. `LVMCluster` is created with a [finalizer][finalizer] and on deletion of CR,
```
Comment (nbalacha) [inline docs/dev-guide/custom-resource-definitions.md:63]: Not required in this file. It is an implementation detail, not a CRD detail.

File: docs/dev-guide/custom-resource-definitions.md [hunk 32]
```diff
@@ -0,0 +1,126 @@
...
+   [entities][reconciler] created by LVMO will be deleted
+
+## 2. LVMVolumeGroup
+
+| Field    | Type                 | Description                                 |
+|----------|----------------------|---------------------------------------------|
+| `spec`   | LVMVolumeGroupSpec   | Desired Volume Groups on respective nodes   |
```
Comment (nbalacha) [inline docs/dev-guide/custom-resource-definitions.md:70]:
```suggestion
| `spec`   | LVMVolumeGroupSpec   | Specification of a single Volume Groups   |
```

File: docs/dev-guide/custom-resource-definitions.md [hunk 33]
```diff
@@ -0,0 +1,126 @@
...
+| `status` | LVMVolumeGroupStatus | Status of Volume Groups on respective nodes |
```
Comment (nbalacha) [inline docs/dev-guide/custom-resource-definitions.md:71]:
```suggestion
| `status` | LVMVolumeGroupStatus | Status of the Volume Group |
```

File: docs/dev-guide/custom-resource-definitions.md [hunk 34]
```diff
@@ -0,0 +1,126 @@
...
+
+### LVMClusterSpec
+
+| Field            | Type                  | Description                           |
+|------------------|-----------------------|---------------------------------------|
+| `deviceSelector` | \*DeviceSelector      | Selector to match devices on the node |
+| `nodeSelector`   | \*corev1.NodeSelector | Selector to match nodes               |
```
Comment (nbalacha) [inline docs/dev-guide/custom-resource-definitions.md:78]:
```suggestion
| `nodeSelector`   | \*corev1.NodeSelector | Selector to determine nodes on which the volume group should be created            |
```

File: docs/dev-guide/custom-resource-definitions.md [hunk 35]
```diff
@@ -0,0 +1,126 @@
...
+### LVMVolumeGroupStatus
+
+- Not Implemented, empty struct
+
+### Lifecycle
+
+1. Created and managed by LVMO, this contains subset of fields from LVMCluster
+   CR. Volume Groups with same name can exist on different nodes and to
+   reconcile based on this CR reduces the burden for multiple controllers
+
```
Comment (nbalacha) [inline docs/dev-guide/custom-resource-definitions.md:89]:
```suggestion
1. The LVMO creates an LVMVolumeGroup CR for each deviceClass in the LVMCluster CR. The LVMVolumeGroups are reconciled by the vgmanager controllers.
```

**Review by nbalacha — CHANGES_REQUESTED** [review]

File: docs/dev-guide/custom-resource-definitions.md (hunk 25 above)
Comment (leelavg) [reply docs/dev-guide/custom-resource-definitions.md:8]: removing this file, same for rest of the comments in this file.

**Comment by leelavg** [comment]
addressed all comments and separated into #99 #100 and #101

**Pull request closed without merging**